lib/versioning/VersionID.js

```diff
@@ -0,0 +1,136 @@
+const LENGTH_TS = 14; // timestamp: epoch in ms
+const LENGTH_SQ = 6; // position in ms slot
+const LENGTH_ST = 5; // site identifier
+
+// empty string template for the variables in a versionId
+const TEMPLATE_TS = new Array(LENGTH_TS + 1).join('0');
+const TEMPLATE_SQ = new Array(LENGTH_SQ + 1).join('0');
+const TEMPLATE_ST = new Array(LENGTH_ST + 1).join(' ');
+
+// site identifier, like PARIS, TOKYO; will be trimmed if exceeding max length
+const SITE_ID = `${process.env.SITE_ID}${TEMPLATE_ST}`.slice(0, LENGTH_ST);
+
+// constants for max epoch and max sequential number in the same epoch
+const MAX_TS = Math.pow(10, LENGTH_TS) - 1; // good until 16 Nov 5138
+const MAX_SQ = Math.pow(10, LENGTH_SQ) - 1; // good for 1 billion ops
+
+// the earliest versionId, used for versions before versioning
+const VID_INF = `${TEMPLATE_TS}${MAX_TS}`.slice(-LENGTH_TS) +
+                `${TEMPLATE_SQ}${MAX_SQ}`.slice(-LENGTH_SQ) + SITE_ID;
+
+// constants for versionId encryption; these values (including the encryption
+// password) should never be reconfigured otherwise existing versionIds will
+// not work; think about using Federation's keyfile as the encryption password
+const VID_CRYPTO_ALGORITHM = process.env.VID_CRYPTO_ALGORITHM || 'aes256';
+const VID_CRYPTO_ENCODING = process.env.VID_CRYPTO_ENCODING || 'base64';
+const VID_CRYPTO_ENCODING_TXT = process.env.VID_CRYPTO_ENCODING_TXT || 'utf8';
+// no need to be secured because users cannot actively create a specific version
+// but better off specifying it if there is a need for additional security layer
+const VID_CRYPTO_PASSWORD = process.env.VID_CRYPTO_PASSWORD || 'hangngon';
+
+// internal state of the module
+let prvts = 0; // epoch of the last versionId
+let prvsq = 0; // sequential number of the last versionId
+
+/**
+ * This function ACTIVELY (wastes CPU cycles and) waits for an amount of time
+ * before returning to the caller. This should not be used frequently.
+ *
+ * @param {Number} span - time to wait in nanoseconds (1/1000000 millisecond)
+ * @return {Undefined} - nothing
+ */
+function wait(span) {
+    function getspan(diff) {
+        return diff[0] * 1e9 + diff[1];
+    }
+    const start = process.hrtime();
+    while (getspan(process.hrtime(start)) < span);
+}
+
+/**
+ * This function returns a "versionId" string indicating the current time as a
+ * combination of the current time in millisecond, the position of the request
+ * in that millisecond, and the identifier of the local site (which could be
+ * datacenter, region, or server depending on the notion of geographics). This
+ * function is stateful which means it keeps some values in the memory and the
+ * next call depends on the previous call.
+ *
+ * @param {string} info - the additional info to ensure uniqueness if desired
+ * @return {string} - the formated versionId string
+ */
+function generateVersionId(info) {
+    // Need to wait for the millisecond slot got "flushed". We wait for
```

Comment:
I'm not sure to understand this bit. I get that you wait for one millisecond, but not why ?
Is it because you need to ensure that it starts fresh on a "new" millisecond, due to the inner workings  relying on millisecond + sequence number ? Are you relying on the fact that Node.JS computation is monothreaded + this mechanism, to make sure that any two instantiations of this class cannot start on the same millisecond ?
If yes, please fix doc to reflect this.